src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -999,13 +1001,38 @@ HeapWord* ShenandoahHeap::allocate_memory(ShenandoahAllocRequest& req) {
     // other threads have already depleted the free storage. In this case, a better
     // strategy is to try again, as long as GC makes progress (or until at least
     // one full GC has completed).
+
     size_t original_count = shenandoah_policy()->full_gc_count();
-    while (result == nullptr
-        && (get_gc_no_progress_count() == 0 || original_count == shenandoah_policy()->full_gc_count())) {
-      control_thread()->handle_alloc_failure(req, true);
-      result = allocate_memory_under_lock(req, in_new_region);
-    }
 
+    // Stop retrying and return nullptr to cause OOMError exception if our allocation failed even after:
+    //   a) We experienced a GC that had good progress, or
+    //   b) We experienced at least one Full GC (whether or not it had good progress)
+    //
+    // TODO: Rather than require a Full GC before throwing OOMError, it might be more appropriate for handle_alloc_failure()
+    //       to trigger a concurrent GLOBAL GC, and throw OOMError if we cannot allocate even after GLOBAL GC has finished.
+    //       There is no "perfect" solution here:
+    //
+    //        1. As currently implemented, there may be a race between multiple allocating threads, both attempting
+    //           to allocate very large objects.  The first thread to retry its allocation might succeed and the second
+    //           thread to retry its allocation might fail (because the first thread consumed the newly available memory).
+    //           So the second thread experiences OOMError even through another GC would have reclaimed the memory it wanted
+    //           to allocate.
+    //        2. A GLOBAL GC won't necessarily reclaim all garbage.  Following a concurrent Generational GLOBAL GC, we may
+    //           need to perform multiple concurrent mixed evacuations in order to reclaim all of the dead memory identified
+    //           by the GLOBAL GC mark.  However, the first evacuation performed by the GLOBAL GC will normally reclaim
+    //           a significant amount of garbage (as guided by garbage first heuristic).  If this is not enough memory
+    //           to satisfy the pending allocation request, we are in "dire straits", and a fail-fast OOMError is probably
+    //           the better remediation than repeated attempts to allocate following repeated GC cycles.
+
+    if (result == nullptr) {
+      while ((result == nullptr) && (original_count == shenandoah_policy()->full_gc_count())) {
+        control_thread()->handle_alloc_failure(req, true);
+        result = allocate_memory_under_lock(req, in_new_region);
+      }
+      if ((result != nullptr) && mode()->is_generational()) {
```

Comment:
I am wondering why we are using a `mode()->is_generational()` filter here.

Would this not work in exactly the same way for the non-generational case as well?

If so, I'd extract this fix out more generally and upstream it separately as its own ticket for fixing the OOM'ing behavior for "no progress OOMs".

Reply:
This would work "equally well" for non-generational mode.  I was reluctant to "change the behavior" of non-gen mode, especially because the problem we're facing seems to only manifest in generational mode.

I don't think it would hurt non-gen mode at all.  But perhaps some rare actual customer of non-gen might actually be counting on the existing behavior...  (?)